src/messaging/ExchangeMgr.cpp

```diff
@@ -99,35 +99,33 @@ CHIP_ERROR ExchangeManager::Shutdown()
     return CHIP_NO_ERROR;
 }
 
-ExchangeContext * ExchangeManager::NewContext(const NodeId & peerNodeId, void * appState)
+ExchangeContext * ExchangeManager::NewContext(const NodeId & peerNodeId, ExchangeContextDelegate * delegate)
 {
-    return AllocContext(mNextExchangeId++, peerNodeId, true, appState);
+    return AllocContext(mNextExchangeId++, peerNodeId, true, delegate);
 }
 
-ExchangeContext * ExchangeManager::FindContext(NodeId peerNodeId, void * appState, bool isInitiator)
+ExchangeContext * ExchangeManager::FindContext(NodeId peerNodeId, ExchangeContextDelegate * delegate, bool isInitiator)
 {
     ExchangeContext * ec = ContextPool;
 
     for (int i = 0; i < CHIP_CONFIG_MAX_EXCHANGE_CONTEXTS; i++, ec++)
     {
-        if (ec->GetReferenceCount() > 0 && ec->GetPeerNodeId() == peerNodeId && ec->GetAppState() == appState &&
+        if (ec->GetReferenceCount() > 0 && ec->GetPeerNodeId() == peerNodeId && ec->GetDelegate() == delegate &&
             ec->IsInitiator() == isInitiator)
             return ec;
     }
 
     return nullptr;
 }
 
-CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeContext::MessageReceiveFunct handler,
```

Comment:
With this API change, we will no longer distinguish the solicited from unsolicited message handing as before. The previous API allows the protocol to register multiple different unsolicited message handlers and even adjust it in between.

Reply:
`Delegate` is an object, it is obviously that we register different delegate object for unsolicited handler and solicited handler. The object itself already knows itself is unsolicited or solicited.

Reply:
No, we can not use different delegate objects for unsolicted handler and solicited handler after this change, since matchingUMH->AppState will be assigned to mDelegate of EC finally. I think we need to adjust the protocol accordingly if they use multiple unsolicited handlers for different messages.

Reply:
Is it possible to use switch/case in the `OnMessageReceived` function to solve this ?

```
switch(msgType)
{
case Type1: ...
case Type2: ...
}
```